### PR TITLE
feat(propagation): always propagate tracestate header

### PR DESCRIPTION
**Pull request by ZStriker19**
**What's it do?**
When the tracer is configured to read & write W3C headers, the tracer will propagate the tracestate values from other vendors so long as the traceparent trace-id matches the inherited trace context trace-id. This way we propagate other vendor's tracestate values even if tracecontext propagation style is not first to match headers upon extraction.

**Tests**
I needed to add in some logic around our extraction tests so that tracecontext could be included.  I then added a few tests to make sure we add the tracestate tag under the correct circumstances. The propagation tests could use some refactoring at some point.

 System-tests cover the rest of the expected behavior: https://github.com/DataDog/system-tests/pull/1795

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/settings/config.py [hunk 1]
```diff
@@ -421,6 +421,8 @@ def __init__(self):
         if propagation_style_inject is not None:
             self._propagation_style_inject = propagation_style_inject
 
+        self._propagation_extract_first = asbool(os.getenv("DD_TRACE_PROPAGATION_EXTRACT_FIRST", False))
```
Comment (emmettbutler) [inline ddtrace/settings/config.py:424]: You might consider adding an entry to `docs/configuration.rst`.

File: ddtrace/propagation/http.py [hunk 2]
```diff
@@ -853,16 +857,45 @@ def my_controller(url, headers):
...
         try:
             normalized_headers = {name.lower(): v for name, v in headers.items()}
 
-            # loop through the extract propagation styles specified in order
-            for prop_style in config._propagation_style_extract:
-                propagator = _PROP_STYLES[prop_style]
-                context = propagator._extract(normalized_headers)  # type: ignore
-                if context is not None:
-                    return context
+            # or tracer configured to extract first only
```
Comment (emmettbutler) [inline ddtrace/propagation/http.py:863]:
```suggestion
            # tracer configured to extract first only
```

File: ddtrace/propagation/http.py [hunk 3]
```diff
@@ -853,16 +857,45 @@ def my_controller(url, headers):
...
+            if config._propagation_extract_first:
+                # loop through the extract propagation styles specified in order, return whatever context we get first
+                for prop_style in config._propagation_style_extract:
+                    propagator = _PROP_STYLES[prop_style]
+                    context = propagator._extract(normalized_headers)  # type: ignore
+                    if context is not None:
+                        return context
+            # loop through all extract propagation styles
```
Comment (emmettbutler) [inline ddtrace/propagation/http.py:871]: You could reduce the complexity and depth of this nested conditional by factoring the branches into functions that do what these comments say.

**Review by emmettbutler — APPROVED** [review]

File: docs/configuration.rst [hunk 4]
```diff
@@ -361,6 +361,17 @@ The following environment variables for the tracer are supported:
      version_added:
        v1.7.0: The ``b3multi`` propagation style was added and ``b3`` was deprecated in favor it.
 
+    DD_TRACE_PROPAGATION_EXTRACT_FIRST:
+     type: Boolean
+     default: False
+     description: |
+        Whether the propagator stops after extracting the first header.
+
+        Example: ``DD_TRACE_PROPAGATION_STYLE_INJECT="datadog,b3multi"`` to inject both ``x-datadog-*`` and ``x-b3-*``
```
Comment (emmettbutler) [inline docs/configuration.rst:370]: This doesn't completely seem like an example of `DD_TRACE_PROPAGATION_EXTRACT_FIRST`.

**Review by emmettbutler — CHANGES_REQUESTED** [review]

File: docs/configuration.rst (hunk 4 above)
Comment (ZStriker19) [reply docs/configuration.rst:370]: Yeah.... this is what I get for running to lunch mid-way through work.

**Review by emmettbutler — APPROVED** [review]

**Pull request merged**